runtime/vm/jvminit.c

```diff
@@ -6191,6 +6197,112 @@ shutDownHookWrapper(struct J9PortLibrary* portLibrary, U_32 gpType, void* gpInfo
+ *
+ * @param vmThread pointer to a J9VMThread
+ * @param signal integer value of the signal
+ *
+ * @return void
+ */
+static void
+signalDispatch(J9VMThread *vmThread, I_32 signal) {
+	J9JavaVM *vm = vmThread->javaVM;
+	J9NameAndSignature nas = {0};
+	I_32 args[] = {signal};
+
+	nas.name = (J9UTF8 *)&j9_dispatch;
+	nas.signature = (J9UTF8 *)&j9_int_void;
+
+	enterVMFromJNI(vmThread);
+
+	if (J2SE_VERSION(vm) >= J2SE_19) {
+		runStaticMethod(vmThread, (U_8*)"jdk/internal/misc/Signal", &nas, 1, (UDATA *)args);
+	} else {
+		runStaticMethod(vmThread, (U_8*)"sun/misc/Signal", &nas, 1, (UDATA *)args);
+	}
+
+	internalExceptionDescribe(vmThread);
+	releaseVMAccess(vmThread);
+}
+
+/* @brief This handler will be invoked by the asynchSignalReporterThread
+ * in omrsignal.c once registered using j9sig_set_*async_signal_handler
+ * for a specific signal.
+ *
+ * @param portLibrary the port library
+ * @param gpType port library signal flag
+ * @param gpInfo GPF information (will be NULL in this case)
+ * @param userData user data (will be a pointer to J9JavaVM in this case)
+ *
+ * @return 0 on success and non-zero on failure
+ *
+ */
+static UDATA
+predefinedHandlerWrapper(struct J9PortLibrary *portLibrary, U_32 gpType, void *gpInfo, void *userData) {
+	J9JavaVM *vm = (J9JavaVM *)userData;
+	J9JavaVMAttachArgs attachArgs = {0};
+	J9VMThread *vmThread = NULL;
+	IDATA result = JNI_ERR;
+	BOOLEAN shutdownStarted = FALSE;
+	I_32 signal = 0;
+	PORT_ACCESS_FROM_JAVAVM(vm);
+
+	signal = j9sig_map_portlib_signal_to_os_signal(gpType);
+	/* Don't invoke handler if signal is 0 or negative, or if -Xrs is specified */
+	if ((signal <= 0) || J9_ARE_ANY_BITS_SET(vm->sigFlags, J9_SIG_XRS | J9_SIG_XRS_ASYNC)) {
+		return 1;
+	}
+
+	/* Don't invoke handler if JVM exit or shutdown has started. */
+	omrthread_monitor_enter(vm->runtimeFlagsMutex);
+	if (J9_ARE_ANY_BITS_SET(vm->runtimeFlags, J9_RUNTIME_EXIT_STARTED | J9_RUNTIME_SHUTDOWN_STARTED)) {
+		shutdownStarted = TRUE;
+	}
+	omrthread_monitor_exit(vm->runtimeFlagsMutex);
+
+	if (shutdownStarted) {
+		return 1;
+	}
+
+	attachArgs.version = JNI_VERSION_1_8;
+	attachArgs.name = "JVM Signal Thread";
+	attachArgs.group = vm->systemThreadGroupRef;
+
+	/* Attach current thread as a daemon thread */
+	result = internalAttachCurrentThread(vm, &vmThread, &attachArgs,
+				J9_PRIVATE_FLAGS_DAEMON_THREAD | J9_PRIVATE_FLAGS_SYSTEM_THREAD | J9_PRIVATE_FLAGS_ATTACHED_THREAD,
+				omrthread_self());
+
+	if (JNI_OK != result) {
+		/* Thread couldn't be attached. So, we can't run Java code. */
+		return 1;
+	}
+
+	/* Run handler (Java code). */
+	signalDispatch(vmThread, signal);
+
+	DetachCurrentThread((JavaVM *)vm);
+
+	return 0;
+}
+
+void *
+registerPredefinedHandler(J9JavaVM *vm, UDATA signal) {
+	void *oldOSHandler = NULL;
+	UDATA portlibSignalFlag = 0;
+
+	PORT_ACCESS_FROM_JAVAVM(vm);
+
+	portlibSignalFlag = j9sig_map_os_signal_to_portlib_signal(signal);
+	if (0 != portlibSignalFlag) {
```

Comment:
Please add a tracepoint for the failure case.

Reply:
something like this?
```
if (0 != portlibSignalFlag) {
    oldOSHandler = j9sig_set_single_async_signal_handler(predefinedHandlerWrapper, vm, portlibSignalFlag);
} else {
    Trc_VM_registerPredefinedHandler_invalidPortlibSignalFlag(portlibSignalFlag);
}
```

Reply:
Yes, using a level 1 TraceException tracepoint.